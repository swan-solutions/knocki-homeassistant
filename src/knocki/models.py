"""Models for Knocki API."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any, Self

from mashumaro import field_options
from mashumaro.mixins.orjson import DataClassORJSONMixin


@dataclass(slots=True)
class TokenResponse:
    """Token response from Knocki API."""

    token: str
    user_id: str

    @classmethod
    def from_api(cls, data: dict[str, Any]) -> Self:
        """Parse token response from API."""
        return cls(
            token=data["data"][0]["attributes"]["id"],
            user_id=data["included"][0]["attributes"]["id"],
        )


@dataclass
class Trigger(DataClassORJSONMixin):
    """Trigger model."""

    device_id: str = field(metadata=field_options(alias="device"))
    details: TriggerDetails


@dataclass
class TriggerDetails(DataClassORJSONMixin):
    """Trigger details model."""

    trigger_id: int = field(metadata=field_options(alias="id"))
    name: str = field(metadata=field_options(alias="name"))


class EventType(StrEnum):
    """Event types for Knocki."""

    CREATED = "actionCreated"
    UPDATED = "actionUpdated"
    DELETED = "actionDeleted"
    TRIGGERED = "actionTriggered"


@dataclass
class Event(DataClassORJSONMixin):
    """Event model."""

    event: EventType
    payload: Trigger
