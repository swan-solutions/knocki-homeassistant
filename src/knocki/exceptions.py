"""Asynchronous Python client for Knocki."""


class KnockiError(Exception):
    """Generic exception."""


class KnockiConnectionError(KnockiError):
    """Knocki connection exception."""
