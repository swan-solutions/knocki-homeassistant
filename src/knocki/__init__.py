"""Asynchronous Python client for Knocki."""

from knocki.exceptions import KnockiConnectionError, KnockiError
from knocki.knocki import KnockiClient
from knocki.models import TokenResponse

__all__ = ["TokenResponse", "KnockiClient", "KnockiConnectionError", "KnockiError"]
