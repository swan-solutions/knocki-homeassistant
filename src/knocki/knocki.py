"""Asynchronous Python client for Knocki."""

from __future__ import annotations

import asyncio
from asyncio import timeout
from dataclasses import dataclass, field
from importlib import metadata
from typing import TYPE_CHECKING, Any, Callable

from aiohttp import ClientSession
from aiohttp.hdrs import METH_DELETE, METH_GET, METH_POST
from mashumaro.codecs.orjson import ORJSONDecoder
import orjson
from yarl import URL

from knocki.exceptions import KnockiConnectionError
from knocki.models import Event, EventType, TokenResponse, Trigger

if TYPE_CHECKING:
    from typing_extensions import Self


VERSION = metadata.version(__package__)

_URLS = {
    True: "staging.knocki.com",
    False: "production.knocki.com",
}

WEB_SOCKET_URL = "wss://ivg6nmkds4.execute-api.us-east-2.amazonaws.com/production"


@dataclass
class KnockiClient:
    """Main class for handling connections with Knocki."""

    token: str | None = None
    session: ClientSession | None = None
    staging: bool = False
    request_timeout: int = 10
    _close_session: bool = False
    _listeners: dict[EventType, list[Callable[[Event], None]]] = field(
        default_factory=dict
    )

    async def _request(
        self,
        uri: str,
        *,
        method: str = METH_GET,
        data: dict[str, Any] | None = None,
    ) -> str:
        """Handle a request to Knocki."""
        url = URL.build(scheme="https", host=_URLS[self.staging]).joinpath(uri)

        headers = {
            "User-Agent": f"PythonKnocki/{VERSION}",
            "Accept": "application/json",
        }

        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    headers=headers,
                    json=data,
                )
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to Knocki"
            raise KnockiConnectionError(msg) from exception

        content_type = response.headers.get("Content-Type", "")
        if response.status != 200 or "application/json" not in content_type:
            text = await response.text()
            msg = "Unexpected response from Knocki"
            raise KnockiConnectionError(
                msg,
                {"Content-Type": content_type, "response": text},
            )

        return await response.text()

    async def login(self, email: str, password: str) -> TokenResponse:
        """Login to Knocki."""
        data = {
            "data": [
                {
                    "type": "tokens",
                    "attributes": {
                        "email": email,
                        "password": password,
                        "type": "auth",
                    },
                }
            ]
        }

        response = await self._request("tokens", method=METH_POST, data=data)

        return TokenResponse.from_api(orjson.loads(response))  # pylint: disable=maybe-no-member

    async def link(self) -> None:
        """Link Knocki account."""
        await self._request("accounts/homeassistant/v1/link", method=METH_POST)

    async def unlink(self) -> None:
        """Unlink Knocki account."""
        await self._request("accounts/homeassistant", method=METH_DELETE)

    async def get_triggers(self) -> list[Trigger]:
        """Get triggers from Knocki."""
        response = await self._request("actions/homeassistant")
        return ORJSONDecoder(list[Trigger]).decode(response)

    async def start_websocket(self) -> None:
        """Start websocket connection."""
        url = WEB_SOCKET_URL + f"?token={self.token}"

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with self.session.ws_connect(url) as ws:
                async for msg in ws:
                    event = Event.from_json(msg.data)  # pylint: disable=maybe-no-member
                    for listener in self._listeners.get(event.event, []):
                        listener(event)
        except Exception as exception:
            err_msg = "Error occurred while connecting to Knocki websocket"
            raise KnockiConnectionError(err_msg) from exception

    def register_listener(
        self, event_type: EventType, listener: Callable[[Event], None]
    ) -> Callable[[], None]:
        """Register a listener."""
        if event_type not in self._listeners:
            self._listeners[event_type] = [listener]
        else:
            self._listeners[event_type].append(listener)
        return lambda: self._listeners[event_type].remove(listener)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The KnockiClient object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
