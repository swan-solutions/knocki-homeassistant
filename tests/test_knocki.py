"""Asynchronous Python client for Knocki."""

from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any

import aiohttp
from aiohttp.hdrs import METH_DELETE, METH_GET, METH_POST
from aioresponses import CallbackResult, aioresponses
import pytest

from knocki import (
    KnockiClient,
    KnockiConnectionError,
    KnockiError,
    KnockiInvalidAuthError,
)

from . import load_fixture
from .const import BASE_URL, HEADERS, UNAUTHORIZED_HEADERS

if TYPE_CHECKING:
    from syrupy import SnapshotAssertion


async def test_putting_in_own_session(
    responses: aioresponses,
) -> None:
    """Test putting in own session."""
    responses.post(
        f"{BASE_URL}/tokens",
        status=200,
        body=load_fixture("tokens.json"),
    )
    async with aiohttp.ClientSession() as session:
        knocki = KnockiClient(session=session)
        await knocki.login("test@test.com", "test")
        assert knocki.session is not None
        assert not knocki.session.closed
        await knocki.close()
        assert not knocki.session.closed


async def test_creating_own_session(
    responses: aioresponses,
) -> None:
    """Test creating own session."""
    responses.post(
        f"{BASE_URL}/tokens",
        status=200,
        body=load_fixture("tokens.json"),
    )
    knocki = KnockiClient()
    await knocki.login("test@test.com", "test")
    assert knocki.session is not None
    assert not knocki.session.closed
    await knocki.close()
    assert knocki.session.closed


async def test_unexpected_server_response(
    responses: aioresponses,
    knocki_client: KnockiClient,
) -> None:
    """Test handling unexpected response."""
    responses.post(
        f"{BASE_URL}/tokens",
        status=200,
        headers={"Content-Type": "plain/text"},
        body="Yes",
    )
    with pytest.raises(KnockiError):
        assert await knocki_client.login("test@test.com", "test")


async def test_timeout(
    responses: aioresponses,
) -> None:
    """Test request timeout."""

    # Faking a timeout by sleeping
    async def response_handler(_: str, **_kwargs: Any) -> CallbackResult:
        """Response handler for this test."""
        await asyncio.sleep(2)
        return CallbackResult(body="Goodmorning!")

    responses.post(
        f"{BASE_URL}/tokens",
        callback=response_handler,
    )
    async with KnockiClient(request_timeout=1) as knocki:
        with pytest.raises(KnockiConnectionError):
            assert await knocki.login("test@test.com", "test")


async def test_login(
    responses: aioresponses, knocki_client: KnockiClient, snapshot: SnapshotAssertion
) -> None:
    """Test logging in."""
    responses.post(
        f"{BASE_URL}/tokens",
        status=200,
        body=load_fixture("tokens.json"),
    )
    assert await knocki_client.login("test@test.com", "test") == snapshot
    responses.assert_called_once_with(
        f"{BASE_URL}/tokens",
        METH_POST,
        headers=UNAUTHORIZED_HEADERS,
        json={
            "data": [
                {
                    "type": "tokens",
                    "attributes": {
                        "email": "test@test.com",
                        "password": "test",
                        "type": "auth",
                    },
                }
            ]
        },
    )


async def test_invalig_login(
    responses: aioresponses, knocki_client: KnockiClient
) -> None:
    """Test logging in."""
    responses.post(
        f"{BASE_URL}/tokens",
        status=200,
        body=load_fixture("invalid_auth.json"),
    )
    with pytest.raises(KnockiInvalidAuthError):
        assert await knocki_client.login("test@test.com", "test")


async def test_link(
    responses: aioresponses, authenticated_client: KnockiClient
) -> None:
    """Test linking."""
    responses.post(f"{BASE_URL}/accounts/homeassistant/v1/link", status=200, body="{}")
    await authenticated_client.link()
    responses.assert_called_once_with(
        f"{BASE_URL}/accounts/homeassistant/v1/link",
        METH_POST,
        headers=HEADERS,
        json=None,
    )


async def test_unlink(
    responses: aioresponses, authenticated_client: KnockiClient
) -> None:
    """Test unlinking."""
    responses.delete(f"{BASE_URL}/accounts/homeassistant", status=200, body="{}")
    await authenticated_client.unlink()
    responses.assert_called_once_with(
        f"{BASE_URL}/accounts/homeassistant",
        METH_DELETE,
        headers=HEADERS,
        json=None,
    )


async def test_get_triggers(
    responses: aioresponses,
    authenticated_client: KnockiClient,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting triggers."""
    responses.get(
        f"{BASE_URL}/actions/homeassistant",
        status=200,
        body=load_fixture("triggers.json"),
    )
    assert await authenticated_client.get_triggers() == snapshot
    responses.assert_called_once_with(
        f"{BASE_URL}/actions/homeassistant", METH_GET, headers=HEADERS, json=None
    )
