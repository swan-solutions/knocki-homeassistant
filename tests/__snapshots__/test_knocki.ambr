# serializer version: 1
# name: test_get_triggers
  list([
    dict({
      'details': dict({
        'name': 'Aaaa',
        'trigger_id': 31,
      }),
      'device_id': 'KNC1-W-00000214',
    }),
  ])
# ---
# name: test_login
  dict({
    'token': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhYWFhYWFhYS04Y2M3LTQzYWEtYmM1ZS05NjhhNzEzYjM0YjIiLCJlbWFpbCI6InRlc3RAdGVzdC5jb20iLCJwd2QiOiJBQTFGNURFNEMwIiwidGYiOmZhbHNlLCJhZ2VudCI6IlBvc3RtYW5SdW50aW1lLzcuMzkuMCIsImlhdCI6MTcxNzU5MzE1Mn0.fxwTBm-JsxWCCTt9QCBQzgK-6sjL0TCeKKtp5_B2iek',
    'user_id': 'aaaaaaaa-8cc7-43aa-bc5e-968a713b34b2',
  })
# ---
